apiserver/logsink.go

```diff
@@ -7,27 +7,105 @@ import (
 	"io"
 	"net/http"
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/juju/errors"
 	"github.com/juju/loggo"
+	"github.com/juju/utils/clock"
 	"github.com/juju/version"
 	"gopkg.in/juju/names.v2"
 
 	"github.com/juju/juju/apiserver/logsink"
 	"github.com/juju/juju/apiserver/params"
 	"github.com/juju/juju/state"
+	"github.com/juju/juju/state/logdb"
+)
+
+const (
+	// dbLoggerBufferSize is the capacity of the log buffer.
+	// When the buffer fills up, it will be flushed to the
+	// database.
+	dbLoggerBufferSize = 1024
```

Comment:
I should note that Bulk has a limit of 1000 records (according to txn pruning) so we probably want to limit the chunk size slightly smaller.